### ISSUE TITLE
[Discussion] findings schema versioning

### ISSUE THREAD

**Issue by olegsu** [elastic/cloudbeat#493]
**Motivation**
As we moved to GA and are planning cloudbeat backward compatibility strategy, we need to make sure that the event schema that cloudbeat sends will be accepted.
As of today, we don't have a way to know in the stack that something was changed/removed/added.
Adding a version to cloudbeat will give us the ability to relay the structure of the data ( that is not part of the ECS ) and if needed to add/remove/update fields in the ingest pipeline.

The end result will be that Kibana will work with a selected (latest) version and will build all the dashboards/ queries and filters based on that.

**Related**
* https://github.com/elastic/kibana/issues/144666
* https://github.com/elastic/cloudbeat/issues/412
* ~~https://github.com/elastic/cloudbeat/issues/496~~
* https://github.com/elastic/security-team/issues/3792

**Comment by olegsu** [comment]
I don't see any [ecs](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) potential field that this field should be set on.
I would suggest adding `scheme.version` to the event as a hardcoded value (`v1`, `v2` ... )

**Comment by eyalkraft** [comment]
Hi @olegsu, very important!

What are some advantages/disadvantages of introducing a new `schema` version separate from the versions of cloudbeat?
maybe this makes https://github.com/elastic/security-team/issues/3792 related

**Comment by kfirpeled** [comment]
> I don't see any [ecs](https://www.elastic.co/guide/en/ecs/current/ecs-field-reference.html) potential field that this field should be set on. I would suggest adding `schema.version` to the event as a hardcoded value (semantic versions `v1.0.0`, `v1.0.1` ... )

I had the this idea, that since our `cloudbeat` is the single producer of findings we can use the `cloudbeat` version to identify findings certain features. wdyt? is it better than introduce now a new field?

**Comment by kfirpeled** [comment]
Good point @eyalkraft, had the same thought when rethinking this issue.

**Comment by olegsu** [comment]
Reading the spec again I see that I missed this 
> The observer.* fields shall be populated with details of the system, if any, that detects, observes and/or creates a network, security, or application **event** or metric
This is a good place to have it then.
Closing https://github.com/elastic/cloudbeat/issues/496 if favor of https://github.com/elastic/security-team/issues/3792

**Comment by oren-zohar** [comment]
@olegsu @kfirpeled @eyalkraft so instead of sending it into `resource.version.cloudbeat.semantic_version` you're suggesting populating `observer.version.cloudbeat.semantic_version`, and so on?

**Comment by kfirpeled** [comment]
correct, the suggestion is to have the versioning under `observer.*`, tho the field name and structure haven't been defined yet

**Comment by eyalkraft** [comment]
@oren-zohar @olegsu @kfirpeled  I don't necessarily think it's a better solution. Maybe it is.
This is why I asked

> What are some advantages/disadvantages of introducing a new `schema` version separate from the versions of cloudbeat?

I'll start:

One disadvantage for adding a new field and a new version is that it might be confusing for users (and developers), taking into consideration that there are already many components with different versions involved (agent, cloudbeat, Kibana, policies?, ECS)

One advantage for adding a new field and a new version is that Kibana could remain independent from cloudbeat versions, and the schema specific code there wouldn't have to change every time cloudbeat version changes, or have any advanced logic to figure out which schema is used based on the cloudbeat version. This makes the schema a declared contract, which cloudbeat, Kibana, and possibly other components can implement.
This can leave the Code in Kibana cleaner since it will check for the schema version and wouldn't have to to any advanced calculations.

What so you all think? What are some more advantages and disadvantages?
Which approach is preferred?

**Comment by olegsu** [comment]
You are right, @eyalkraft.
Cloudbeat and Kibana coupling is not a good idea.
Instead, Kibana (and other components) should be able to handle different versions of the data (findings), or filter always latest version if a transformation is possible.

**Comment by eyalkraft** [comment]
> Another thing that I think is worth considering is sending the data to an index with an explicit version.

@olegsu super interesting!

**Comment by kfirpeled** [comment]
hmm I see what you're saying. I tend to agree.
So `findings` versioning is out of scope. I still find it useful to have versions of the cloudbeat, policies and api-server attached for troubleshooting purposes.
I'd make sure that kibana doesn't relay on these fields

does that make sense?

**Comment by olegsu** [comment]
Agree, for troubleshooting we need to have it.

Talked @CohenIdo about [this](https://github.com/elastic/cloudbeat/issues/493#issuecomment-1323243493) suggestion. 
It seems that as long as we have the `_latest` index we might be needed to add `_latest` with the same version. This need to be verified though.
More things that we need to consider/research before starting POC
1. What is the approach that we need to take when (if possible) a stack upgrade will bump more than one version. For example, from 8.4 to 8.6.
2. `_latest` index should also be versioned

**Comment by olegsu** [comment]
1. There is no common approach to running migration ( only for `saved_object` do we have it). We can achieve this using the [painless script](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-scripting-painless.html). 
2. We can use `Index Aliases` and use the alias name to query from Kibana, so that when no `_reindex` is required we need only to update the target index.

**Comment by olegsu** [comment]
The next step will be to run POC that will map all the technical gaps and requirements, and also get a sense of time estimation.
